### ISSUE TITLE
orgzly keeps syncing to a conflict branch even after the branch has been merged

### ISSUE THREAD

**Issue by wilderjds** [orgzly-revived/orgzly-android-revived#237]
- [x] I have searched for existing issues that may be the same as or related to mine.

Please complete the following information:

    Device: Pixel 7
    Android Version: 14 
    Orgzly Revived Version: 
    Repository type(s) used: git 

On a git repository, orgzly keep syncing to a merge-conflict branch even if the branch has been merged to master. No conflicts are present; I even removed the branch from the remote, but orgzly keeps creating it. 
This has worked OK for some time but with the last conflict it is not resetting to master.

**Comment by amberin** [comment]
I am currently testing a completely new git sync logic, so my memory of how the current logic works may be wrong, but I believe there is a bug which means that if you remove the conflict branch from the remote, this will happen, even if the main branch has been updated. I think you need to resolve the conflict on the temp branch, and then leave it like that on the remote side. Otherwise, Orgzly will simply push its view of the conflict branch, and not attempt to return to the main branch. (Once you're back on main, you can safely delete the temp branch.)

This is of course confusing and bad, but it is solved in my ongoing overhaul of the git sync logic.

Please try this and let me know if it works.

**Comment by wilderjds** [comment]
Thanks for your reply.  It does not work as it should; here is what happens

I merge the conflict branch to master on my laptop, then push to the remote
After that, I sync on Android
Then I push some changes from the laptop to master
After that, I sync on Android; changes are picked up, but the notebook is still on the same conflict branch
If I make some changes from Android, they go to the same conflict branch, and not to master

**Comment by amberin** [comment]
Sounds very strange indeed... If you look at the repo settings, is the main branch name correct? (See #29.)

**Comment by amberin** [comment]
FYI, this is where you should end up in the code, but for some reason don't (or attemptReturnToMainBrach fails): https://github.com/orgzly-revived/orgzly-android-revived/blob/master/app/src/main/java/com/orgzly/android/git/GitFileSynchronizer.java#L318

But, as I was trying to say earlier, one known issue is that, if the current branch does not exist on the remote (because you have deleted the temp branch), you will end up here, pushing the current branch instead of returning back to main: https://github.com/orgzly-revived/orgzly-android-revived/blob/master/app/src/main/java/com/orgzly/android/git/GitFileSynchronizer.java#L327

**Comment by wilderjds** [comment]
> When you say you are receiving new changes on master, does the book status say "loaded from branch master"? Because that message means that the repo is on that branch. Any subsequent push of a temp branch should be a new one.

No, 
I get  "loaded from branch merge-conflict-xxx", but the changes we re only pushed to master

**Comment by wilderjds** [comment]
> Sounds very strange indeed... If you look at the repo settings, is the main branch name correct? (See #29.)

Yes, the branch is "master". Always had been.

**Comment by stfl** [comment]
```
5422c66 *   origin/master origin/merge-freelance.org-28ae343-2024-05-02_142716 origin/merge-freelance.org-28ae343-2024-05-02_140728 Merge remote-tracking branch 'origin/merge-freelance.org-28ae343-2024-05-02_140728'
        |\  
a07f3a3 | * Orgzly update: inbox.org
955a6b4 * | origin/merge-freelance.org-28ae343-2024-05-02_140123 changes from amsel on Thu May 2 02:06:53 PM CEST 2024
78304c9 * | changes from amsel on Thu May 2 02:06:48 PM CEST 2024
        |/  
4d3d93a * Orgzly update: inbox.org
e221d7e * Orgzly update: inbox.org
76640f8 * origin/merge-freelance.org-28ae343-2024-05-02_084944 changes from amsel on Thu May 2 01:19:29 PM CEST 2024
```

**Comment by stfl** [comment]
Yes. That makes sense. I am the end doing a rebase forces the other end to do a reset.

I am using git-sync for my other devices which rebases as the default sync procedure.

**Comment by amberin** [comment]
@wilderjds Interesting. I'd love to know how your master branch ended up in that diverged state...

**Comment by wilderjds** [comment]
> @wilderjds Interesting. I'd love to know how your master branch ended up in that diverged state...

No idea tbh, but will report here if it happens again. Now that I found out that I can keep track of stuff in termux, it will be easier to see what is going on.

**Comment by bkhl** [comment]
I have this issue too and had a couple of suggestions that would have made the current situation less confusing:

* show in the _Notebooks_ pane which branch a particular notebook is currently synced to.
* when triggering a _force load_ from the repo, take that to mean we want to reset to the main branch and forget about any conflicts or other branches.

@amberin of course if you're working on some other merge logic that will make this unnecessary that's even better, but just leaving it here for what it is.

**Comment by amberin** [comment]
>     * show in the _Notebooks_ pane which branch a particular notebook is currently synced to.

The challenge with that is that all notebook detail fields are generic and common to all repo types. This is why we currently show it in the "last action" status string. But it really affects all notebooks from the same repository - we switch branches in the work tree.

>     * when triggering a _force load_ from the repo, take that to mean we want to reset to the main branch and forget about any conflicts or other branches.

I agree that this would probably be desirable. Problem is, again, that the UI is common to all repo types, so how to make this clear to the user? Forcd-load is basically an interface method which all repos implement differently, but outside of that call, repos should look the same.

>  @amberin of course if you're working on some other merge logic that will make this unnecessary that's even better

Yes, me and @chaoflow are currently testing his suggested workflow where we never switch branches, but we just force-push the last commit to a predefined remote branch whenever there is a conflict. And always attempt rebase if our head and the remote main branch head do not match. It feels good.

The big challenge for me is how this new sync workflow should best relate to other repo types, because in this branch, git repos are synced as one unit, instead of syncing all books separately, regardless of repo type, which Orgzly has always done so far.

**Comment by amberin** [comment]
> I have this issue too

@bkhl Do you know a way to reproduce it? If so, it would be nice to fix the bug in the current code.

**Comment by bkhl** [comment]
> I agree that this would probably be desirable. Problem is, again, that the UI is common to all repo types, so how to make this clear to the user? Forcd-load is basically an interface method which all repos implement differently, but outside of that call, repos should look the same.

I would have said this is already clear since the settings for the repo includes which branch you sync to. This is why I at first assumed that doing a force load would forget about any local changes and remove all conflicts.

> Do you know a way to reproduce it? If so, it would be nice to fix the bug in the current code.

Not a step-by-step, but I could try making some using a public repo and some device that I don't already have Orgzly on. It happens fairly frequently so hopefully it should be easy.

I'll try to do it soon, but no delivery date promises.

I'll say though that I've only rarely seen actual conflicts, mostly it's ones that are handled automatically if i merge or rebase on my computer using the Git CLI, so I would expect that the new workflow you are working on would solve most of our issues. (Assuming the Git Library used in Orgzly does automatic conflict resolution as well as Git itself).

**Comment by amberin** [comment]
> I would have said this is already clear since the settings for the repo includes which branch you sync to. This is why I at first assumed that doing a force load would forget about any local changes and remove all conflicts.

Yes, you are right - it should be the least surprising behavior: throw away all local state.

>  It happens fairly frequently so hopefully it should be easy

Hmm, sounds like it happens every time you run into a merge conflict?

I wonder if it's something about merging to the main branch. I usually merge the main branch into the conflict branch, push it to remote, and then trigger sync in Orgzly. Doing it the other way is supposed to work, but there may be bugs I'm not aware of.

**Comment by bkhl** [comment]
> Hmm, sounds like it happens every time you run into a merge conflict?

It's usually not a "conflict" in Git lingo, but rather just that there's been edits on both sides. In other words, when I later to "git merge" on my computer it's merged without conflict.

So, it's more like it happens any time I make edits on both sides. Sometimes I could also swear it happens when I haven't edited anything in Orgzly, but it could maybe be that it does some reformatting in some cases so might want to push up "non-significant" changes. Hopefully I can document this a bit better when I do some more structured experimentation.

> I usually merge the main branch into the conflict branch, push it to remote, and then trigger sync in Orgzly. Doing it the other way is supposed to work, but there may be bugs I'm not aware of.

That's what I've done too after finding this thread. It sometimes makes Orgzly calm down, sometimes it just keeps pushing to the `merge-*` branch again. This is why I suggested showing in the UI which branch it's on, so at least I don't have to worry about it when it's on the main branch.

**Comment by amberin** [comment]
> This is why I suggested showing in the UI which branch it's on

Yeah, that would be nice. All we have now is the"last action" status string showing which branch we last wrote to. That branch will be used by all notebooks - the whole work tree is on that branch until we return to main.

**Comment by amberin** [comment]
If the issue arises when without merge conflicts, then it should like a severe bug, so I look forward to getting more details. 👍

**Comment by bkhl** [comment]
Just an update. I haven't had the energy to try to reproduce this from a "clean" repo, but I did try this instead:

I removed the syncing in Orgzly on my phone, deleted all the "merge" branches hanging around in my repo, then set up syncing again from scratch.

After that I've done the merging only from main to the "merge" branches as described above.

This feels a bit backwards of course, but since then I've not seen any "no actual conflicts" conflicts popping up, only "real" conflicts.

So while it's just one data point, maybe having "wrong way" merges in the history of my git repo is what somehow triggered the weird behavior further down the line?

Anyway, maybe it will help someone to know that seemingly I could sort out my most immediate issues by resetting the syncing in Orgzly and then only using the workflow described in the first comment from @amberin in this issue.

**Comment by amberin** [comment]
@bkhl Thank you, I think you are right that this might help others. Especially since this workflow is indeed not very intuitive, and should definitely be considered a bug in the current code. It can surely be fixed, but I haven't put effort into it.

I've been working on an overhaul of the git sync logic where this is no longer an issue. There, Orgzly only ever pushes to a single "conflict" branch, and only fetches changes from the main branch. So you just merge into main, and no more cleaning up merge branches. But I have been busy with other stuff, and some tricky code design questions have arisen which are slowing me down.

But until that gets released, thanks again for helping clarify the current, counter-intuitive conflict resolution behavior.

**Comment by bkhl** [comment]
So I've kept using this workflow and it's mostly fine, though I've noticed there's one thing that's the most common reason for me to get sync conflicts that could maybe be addressed.

Often when I've changed things in Emacs and pushed it, next time Orgzly syncs it will do that and then push a commit after reformatting.

Would it be possible to get it to skip pushing whitespace-only changes like this?
